internal/objectstore/remote/retriever.go

```diff
@@ -0,0 +1,392 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package remote
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync/atomic"
+
+	"github.com/juju/clock"
+	jujuerrors "github.com/juju/errors"
+	"github.com/juju/worker/v4"
+	"github.com/juju/worker/v4/catacomb"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/api"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/errors"
+	"github.com/juju/juju/internal/s3client"
+	internalworker "github.com/juju/juju/internal/worker"
+	"github.com/juju/juju/internal/worker/apiremotecaller"
+)
+
+const (
+	// NoRemoteConnections is returned when there are no remote connections
+	// available.
+	NoRemoteConnections = errors.ConstError("no remote connections available")
+
+	// BlobNotFound is returned when the requested blob is not found on any of
+	// the remote connections.
+	BlobNotFound = errors.ConstError("blob not found")
+)
+
+// BlobsClient is an interface for retrieving objects from an object store.
+type BlobsClient interface {
+	// GetObject returns a reader for the object with the given key in the
+	// given bucket.
+	GetObject(ctx context.Context, bucket, key string) (io.ReadCloser, int64, error)
+}
+
+// NewObjectClientFunc is a function that creates a new BlobsClient.
+type NewObjectClientFunc func(url string, client s3client.HTTPClient, logger logger.Logger) (BlobsClient, error)
+
+// BlobRetriever is responsible for retrieving blobs from remote API servers.
+type BlobRetriever struct {
+	catacomb catacomb.Catacomb
+
+	namespace string
+
+	apiRemoteCallers apiremotecaller.APIRemoteCallers
+	newObjectClient  NewObjectClientFunc
+
+	runner *worker.Runner
+
+	clock  clock.Clock
+	logger logger.Logger
+
+	index uint64
+}
+
+// NewBlobRetriever creates a new BlobRetriever.
+func NewBlobRetriever(apiRemoteCallers apiremotecaller.APIRemoteCallers, namespace string, newObjectClient NewObjectClientFunc, clock clock.Clock, logger logger.Logger) (*BlobRetriever, error) {
+	w := &BlobRetriever{
+		namespace:        namespace,
+		newObjectClient:  newObjectClient,
+		apiRemoteCallers: apiRemoteCallers,
+		clock:            clock,
+		logger:           logger,
+
+		runner: worker.NewRunner(worker.RunnerParams{
+			IsFatal: func(err error) bool {
+				return false
+			},
+			ShouldRestart: func(err error) bool {
+				return false
+			},
+			Clock:  clock,
+			Logger: internalworker.WrapLogger(logger),
+		}),
+	}
+
+	if err := catacomb.Invoke(catacomb.Plan{
+		Site: &w.catacomb,
+		Work: w.loop,
+		Init: []worker.Worker{w.runner},
+	}); err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+// Retrieve returns a reader for the blob with the given SHA256.
+func (r *BlobRetriever) Retrieve(ctx context.Context, sha256 string) (_ io.ReadCloser, _ int64, err error) {
```

Comment:
I know I'm being a pain (and we discussed methods length today), but this method is way too long. Maybe we can split it a little bit.